src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
+    val matchedDisk = matchedResources.filter(_.getName == "disk").map(c => BigDecimal(c.getScalar.getValue)).sum
+    val matchedGpus = matchedResources.filter(_.getName == "gpus").map(c => BigDecimal(c.getScalar.getValue)).sum
+
+    assert(cpus <= matchedCpus)
+    assert(mem <= matchedMem)
+    assert(disk <= matchedDisk)
+    assert(gpus <= matchedGpus)
+  }
+
+  // We have list of (BigDecimal, Resource) pairs in order to avoid using == and != operators with doubles.
+  private case class Resources(
+      cpus: List[(BigDecimal, mesos.Resource)],
+      mem: List[(BigDecimal, mesos.Resource)],
+      disk: List[(BigDecimal, mesos.Resource)],
+      gpus: List[(BigDecimal, mesos.Resource)])
+
+  private class ResourceConsumer(initial: Resources) {
+    private var cpus = initial.cpus
+    private var mem = initial.mem
+    private var disk = initial.disk
+    private var gpus = initial.gpus
+
+    // Find a CPU resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeCpus(quantity: Double): Option[mesos.Resource] = {
+      consume(cpus, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          cpus = resources
+          resource
+      }
+    }
+
+    // Find a MEM resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeMem(quantity: Double): Option[mesos.Resource] = {
+      consume(mem, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          mem = resources
+          resource
+      }
+    }
+
+    // Find a DISK resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeDisk(quantity: Double): Option[mesos.Resource] = {
+      consume(disk, BigDecimal(quantity)) map {
+        case (resources, resource) =>
+          disk = resources
+          resource
+      }
+    }
+
+    // Find a GPU resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeGpus(quantity: Double): Option[mesos.Resource] = {
+      consume(gpus, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          gpus = resources
+          resource
+      }
+    }
+
+    private def consume(
+      resources: List[(BigDecimal, mesos.Resource)],
+      quantity: BigDecimal): Option[(List[(BigDecimal, mesos.Resource)], mesos.Resource)] =
+      if (quantity <= BigDecimal(0)) {
+        None
+      } else {
+        resources.partition(_._1 != quantity) match {
+          case (_, Nil) =>
+            throw new IllegalStateException("failed to find a resource with the given quantity")
+          case (left, pair :: right) =>
+            // return a new resource list not containing the found resource and the found resource itself
+            Some((left ::: right, pair._2))
+        }
+      }
+  }
+
+  private[this] def binPackResources(pod: PodDefinition, matchedResources: Seq[mesos.Resource]): Resources = {
```

Comment:
Are the resources from one offer only? Why is it a sequence?

Reply:
Mesos folks said that an offer can have multiple resources of one kind (say, `disk`).

Reply:
Ah, thanks for clarifying.